vxsandbox/resources/kv.py

```diff
@@ -235,3 +234,187 @@ def handle_incr(self, api, command):
+        `False` otherwise.
+        """
+        return self.status == self.SCANNING
+
+    def saving(self):
+        """
+        Return `True` if the reconciliation is write sandbox key counts
+        back to Redis, `False` otherwise.
+        """
+        return self.status == self.SAVING
+
+
+class ReconciliationLock(object):
+    def __init__(self, redis, expiry):
+        self._redis = redis
+        self._expiry = expiry
+
+    def acquire(self):
+        pass
+
+    def release(self):
+        pass
+
+
+class Reconciler(object):
+    """
+    A task for reconciling key counts.
+
+    Key counts are tracked accurately as keys are added or deleted,
+    but keys that expire in Redis are never removed from a sandboxes
+    key limit.
+
+    :param TxRedisManager redis:
+        Redis connection manager.
+
+    :param int period:
+        Seconds between reconciliation checks. Default is 600s.
+
+    :param int recon_expiry:
+        Seconds before a reconciliation is considered outdated.
+        Default is one day.
+    """
+
+    DEFAULT_PERIOD = 10 * 60  # ten minutes
+    DEFAULT_RECON_EXPIRY = 24 * 60 * 60  # one day
+
+    work_fraction = 0.95  # fraction of the period to do work for
+
+    def __init__(self, redis, period=None, recon_expiry=None):
+        self._redis = redis
+        self._period = period or self.DEFAULT_PERIOD
+        self._recon_expiry = recon_expiry or self.DEFAULT_RECON_EXPIRY
+        self._task = LoopingCall(self.attempt_reconciliation)
+        self._done = None
+
+    def _recon_key(self, *parts):
+        return "#".join(["recon"] + parts)
+
+    def _lock_key(self):
+        return self._recon_key("lock")
+
+    def start(self):
+        """
+        Start attempting reconciliation.
+        """
+        if self._done is None:
+            self._done = self._task.start(self._period, now=False)
+            self._done.addErrback(
+                lambda failure: log.err(
+                    failure, "Reconciliation task failed."))
+
+    @inlineCallbacks
+    def stop(self):
+        """
+        Stop attempting reconciliation.
+        """
+        if self._done is not None:
+            self._task.stop()
+            yield self._done
+            self._done = None
+
+    @inlineCallbacks
+    def attempt_reconciliation(self):
+        """
+        Attempt to perform some reconciliation work.
+
+        Only one worker performs reconciliation work at a time. This
+        reduces load on Redis and simplifies the implementation.
+
+        Work consists off:
+
+        * Load the reconciliation status from Redis.
+        * If no reconciliation is in progress and the last one was recent
+          enough, no work is needed.
+        * If there is a reconciliation in progress, continue scanning
+          keys from where the work stopped, updating counts as needed.
+        * After doing some work, write the results back to the reconciliation
+          status in Redis.
```

Comment:
I know this isn't yet ready for review, but its a bit unclear to me what part the locks play in these steps. Could you explain?